### DISCUSSION TITLE
networkx counterpart for  S(tanford)NAP's  GenRndPowerLaw function

### DISCUSSION THREAD

**Discussion by vaswin0** [networkx/networkx#7525] [Ideas]
I couldn't find the networkx version of SNAP's **GenRndPowerLaw**.

SNAP's **GenRndPowerLaw** function "generates a random scale-free graph with power-law degree distribution with exponent PowerExp. The method uses either the Configuration model  or the Edge Rewiring method ."

If nx doesn't have such a function I would like to propose the following:

From the source  [here](https://github.com/snap-stanford/snap/blob/6924a035aabd1ce0a547b94e995e142f29eb5040/snap-core/ggen.cpp#L34) GenRndPowerLaw does this in two steps:

1. generates a power-law degree sequence
2. passes the degree sequence to the function, configuration model

I couldn't find an nx function for the step1, and obviously, nx has [configuration model](https://networkx.org/documentation/stable/reference/generated/networkx.generators.degree_seq.configuration_model.html).

Is it possible to put these together to make an nx's version of GenRndPowerLaw?

**Comment by dschult** [comment]
The [powerlaw_sequence function](https://networkx.org/documentation/stable/reference/generated/networkx.utils.random_sequence.powerlaw_sequence.html) is in `nx.utils` and I think it does the trick.

**Reply by vaswin0** [reply]
Since nx has both components needed to generate a random scale-free graph with power-law degree distribution with a given exponent, would you consider adding GenRndPowerLaw as a feature enhancement to nx's graph generation methods? I would like to work on it.

**Reply by dschult** [reply]
I'd rather that we not add a new function since the lego-style building blocks are available, the user can put them together in short order.

**Reply by vaswin0** [reply]
I just checked [powerlaw_sequence](https://networkx.org/documentation/stable/reference/generated/networkx.utils.random_sequence.powerlaw_sequence.html). It seems like it is not returning a ~~degree distribution~~ graphical sequence.  Instead, it returns a sequence of real samples from a PL distribution.

**Reply by dschult** [reply]
Well, it can't return a distribution, but it returns a powerlaw sequence.:)

The sequence might or might not be a graphical degree sequence depending on what you require of your degree sequence (are multigraphs allowed? self-loops? Is this an in-degree sequence or out-degree, etc). So in your case, you should probably use `powerlaw_sequence()` inside a loop that checks if it is whatever type of "graphical" you need and repeats if it is not.

This is part of the advantage of the lego-style approach -- there are so many options for how to create a powerlaw graph that making them all accessible from a single function is tricky. The SNAP function choose what it thinks you want if I understand correctly, but that means it doesn't work for lots of other cases and we want to provide those other cases too.

**Reply by rossbar** [reply]
This seems to me like it might make a good docstring and/or gallery example!
